drivers/clock_control/clock_stm32_ll_h7.c

```diff
@@ -231,32 +321,142 @@ static int stm32_clock_control_init(struct device *dev)
 	while (LL_RCC_HSE_IsReady() != 1) {
 	}
 
-	/* Set FLASH latency */
-	LL_FLASH_SetLatency(LL_FLASH_LATENCY_4);
-
 	/* Main PLL configuration and activation */
 	LL_RCC_PLL_SetSource(LL_RCC_PLLSOURCE_HSE);
+
+	pllsrc_clock = CONFIG_CLOCK_STM32_HSE_CLOCK;
+
+#elif defined(CONFIG_CLOCK_STM32_PLL_SRC_CSI)
+	/* Support for CSI oscillator */
+
+	LL_RCC_CSI_Enable();
+	while (LL_RCC_CSI_IsReady() != 1) {
+	}
+
+	/* Main PLL configuration and activation */
+	LL_RCC_PLL_SetSource(LL_RCC_PLLSOURCE_CSI);
+
+	pllsrc_clock = CSI_VALUE;
+
+#elif defined(CONFIG_CLOCK_STM32_PLL_SRC_HSI)
+	/* By default choose HSI as PLL clock source */
+
+	/* @TODO for USB support developer needs to force PLL must be clock from HSE */
+
+	/* Enable HSI oscillator */
+	LL_RCC_HSI_Enable();
+	while (LL_RCC_HSI_IsReady() != 1) {
+	}
+
+	/* Calibrate the HSI */
+	LL_RCC_HSI_SetCalibTrimming(32);
+	/* @TODO make HSI divider configurable */
+	LL_RCC_HSI_SetDivider(LL_RCC_HSI_DIV1);
+
+	/* Main PLL configuration and activation */
+	LL_RCC_PLL_SetSource(LL_RCC_PLLSOURCE_HSI);
+
+	pllsrc_clock = HSI_VALUE;
+
 #else
-	#error "CONFIG_CLOCK_STM32_PLL_SRC_HSE not selected"
-#endif /* CONFIG_CLOCK_STM32_PLL_SRC_HSE */
+
+	/* No clock source selected for PLL, by default, disable the PLL */
+	LL_RCC_PLL_SetSource(LL_RCC_PLLSOURCE_NONE);
+#endif
+
+	/* Configure the PLL dividers/multipliers only if PLL source is configured */
+#if defined(CONFIG_CLOCK_STM32_PLL_SRC_HSE) || \
+	defined(CONFIG_CLOCK_STM32_PLL_SRC_HSI) || \
+	defined(CONFIG_CLOCK_STM32_PLL_SRC_CSI)
+
+
+	vco_input_range = get_vco_input_range(
+			pllsrc_clock,
+			CONFIG_CLOCK_STM32_PLL_M_DIVISOR);
+
 
 	/* Configure PLL1 */
+	/* According to the RM0433 datasheet */
+	/* Select clock source */
+	/* Init pre divider DIVM */
+	LL_RCC_PLL1_SetM(CONFIG_CLOCK_STM32_PLL_M_DIVISOR);
+	/* Config PLL */
+
+	/* VCO sel, VCO range */
+	LL_RCC_PLL1_SetVCOInputRange(vco_input_range);
+	LL_RCC_PLL1_SetVCOOutputRange(LL_RCC_PLLVCORANGE_WIDE);
```

Comment:
In case of vco_input_range between  1MHz to 2MHz, LL_RCC_PLLVCORANGE_MEDIUM should be used.
```
The PLLs integrated into the RCC are completely independent. They offer the following
features:
• Two embedded VCOs:
    – A wide-range VCO (VCOH)
    – A low-frequency VCO (VCOL)
• Input frequency range:
    – 1 to 2 MHz when VCOL is used
    – 2 to 16 MHz when VCOH is used
```

Reply:
I didn't understand it this way.
The choice of VCOH or VCOL depends on the output frequency of the PLL right ?

RM0433: Figure 47. PLL block diagram
VCOH: 192 to 836 MHz
VCOL: 150 to 420 MHz

Input frequency range:
– 1 to 2 MHz **when VCOL** is used
– 2 to 16 MHz **when VCOH** is used

I don't know if this is this statement works both ways. Depending on your VCO choice, you range of frequencies must be 1 to 2 for the VCOL and 2 to 16 for the VCOH.

You might be right on this one and we should probably just set the VCORANGE based on the VCOINPUTRANGE.

Reply:
looking at figure below it works both ways (RM0399)
![image](https://user-images.githubusercontent.com/50730894/88025050-002e6580-cb34-11ea-84d4-6c61b191d3b3.png)

Reply:
I trust you on this one. I have added the necessary code to determine the VOCL/H based on the frequency input range. It works on my H743ZI board so you must be right ! You can have a look at it 😄